### PR TITLE
Introduce kvm-libvirt platform

### PR DESCRIPTION
**Pull request by pothos**
This documents how Lokomotive can be set up with nodes as VM on KVM through the libvirt Terraform provider. It provides a controller and a worker Terraform module to start with.

It provides a quick and easy way to test Lokomotive with a defined Flatcar image and doesn't require to have access to one of the cloud providers or spare hardware.

### Testing

Please test it with the steps mentioned in the documentation.

Depends on https://github.com/kinvolk/terraform-render-bootkube/pull/12

### CODE REVIEW

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf [hunk 1]
```diff
@@ -0,0 +1,96 @@
+resource "libvirt_pool" "controllerstmp" {
+  name = "controllerstmp"
```
Comment (invidian) [inline kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: Hm, it failed for me saying this pool exists. Maybe we should use different pool name to make sure it does not exist in existing setups?

File: kvm-libvirt/flatcar-linux/kubernetes/bootkube.tf [hunk 2]
```diff
@@ -0,0 +1,33 @@
+module "bootkube" {
+  source = "github.com/kinvolk/terraform-render-bootkube?ref=f46efe1427f8d35c01da64fde95a798778cbe31f"  # currently in PR #12
```
Comment (pothos) [inline kvm-libvirt/flatcar-linux/kubernetes/bootkube.tf]: Need to update the `ref` for https://github.com/kinvolk/terraform-render-bootkube/pull/12

Comment (invidian) [reply kvm-libvirt/flatcar-linux/kubernetes/bootkube.tf]: Please also use `git::https`

File: docs/flatcar-linux/kvm-libvirt.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
+# KVM libvirt
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md:1]: Actually there are providers available for both Vagrant and VirtualBox, we could use some of these as well if we care about portability (as @johananl mentioned).

File: docs/flatcar-linux/kvm-libvirt.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+* free RAM (minimum 4 GB)
+* libvirtd and a user who is allowed to create virtual machines
+* Terraform v0.11.x, [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct), and [terraform-provider-libvirt](https://github.com/dmacvicar/terraform-provider-libvirt) installed locally
+
+## libvirtd and User Setup
+
+Ensure that libvirt is set up and accessible for the current user:
+
+```sh
+$ sudo dnf install libvirt-daemon libvirt-devel  # here for Fedora
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: I think we could skip this step and refer to "installation on your platform".

File: docs/flatcar-linux/kvm-libvirt.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+## Cluster
+
+Define a Kubernetes cluster using the module [kvm-libvirt/flatcar-linux/kubernetes](https://github.com/kinvolk/lokomotive-kubernetes/tree/master/kvm-libvirt/flatcar-linux/kubernetes).
+
+Create the `mycluster.tf` file:
+
+```tf
+module "controller" {
+  source = "git::https://github.com/kinvolk/lokomotive-kubernetes/kvm-libvirt/flatcar-linux/kubernetes?ref=<hash>"
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: There should be double slash after `lokomotive-kubernetes`. It's Terraform syntax for module source from git repository subdirectory. https://www.terraform.io/docs/modules/sources.html#modules-in-package-sub-directories

File: docs/flatcar-linux/kvm-libvirt.md [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+  providers = {
+    local    = "local.default"
+    null     = "null.default"
+    template = "template.default"
+    tls      = "tls.default"
+    libvirt  = "libvirt.default"
+  }
+
+  os_image_unpacked = "file:///home/CHANGEME/flatcar_production_qemu_image.img"
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: Maybe `{path.cwd}` ?

File: kvm-libvirt/flatcar-linux/kubernetes/bootkube.tf (hunk 2 above)
Comment (invidian) [inline kvm-libvirt/flatcar-linux/kubernetes/bootkube.tf]: Ditto regarding source url.

**Review by invidian — CHANGES_REQUESTED** [review]
Just a few nits in the comments. I tested it and it works fine. It's actually cool how fast it deploys.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 6 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: I think it's better to ask for a change here, confirming that the image is at the correct path.

**Comment by pothos** [comment]
Thanks for the comments!

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 1 above)
Comment (pothos) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: It uses a random variable now.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 3 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md:1]: Ok, good to know for later.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 5 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Thanks!

**Comment by chrskhl** [comment]
Is this something we could use for the CI? We currently have another VM setup for that.

**Comment by chrskhl** [comment]
I just noticed that @squeed did something like this for Typhoon in https://github.com/poseidon/typhoon/pull/266

Is this based on that?

**Comment by invidian** [comment]
>Is this something we could use for the CI? We currently have another VM setup for that.

That doesn't use Matchbox and PXE, so I doubt in can be used for CI. We should test it on CI later on though.

**Comment by pothos** [comment]
> Is this based on that?

@invidian Pointed me to this as well but the structure is different from the packet/aws files which I used as a basis. In the linked PR the controllers and workers get names instead of the count as done in packet/aws. I took the same mechanism to assign stable IPs to the controllers (starting with .10) and having a machine domain variable but the rest is different, e.g., this PR here does not change the host DNS server for IP resolution but uses node IPs in the configuration, and also uses a temporary disk image pool instead the default one.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 7]
```diff
@@ -0,0 +1,297 @@
+# KVM libvirt
+
+In this tutorial, we'll create a Kubernetes cluster with Flatcar Linux inside VMs on our local machine.
+
+We'll declare the cluster using the Lokomotive Terraform module for the KVM libvirt platform.
+
+Controllers are provisioned to run an `etcd-member` peer and a `kubelet` service. Workers run just a `kubelet` service. A one-time [bootkube](https://github.com/kubernetes-incubator/bootkube) bootstrap schedules the `apiserver`, `scheduler`, `controller-manager`, and `coredns` on controllers and schedules `kube-proxy` and `calico` (or `flannel`) on every node. A generated `kubeconfig` provides `kubectl` access to the cluster.
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: Can we break this long line?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+## Requirements
+
+* free RAM (minimum 4 GB)
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: I would rephrease to "At least 4 GB of free RAM". In any case the first letter in the sentence should be capitalized, unless it is a special term which is normally written in lowercase (e.g. `libvirt`).

File: docs/flatcar-linux/kvm-libvirt.md [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+* libvirtd and a user who is allowed to create virtual machines
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: I would move the permissions to a separate line and explain more explicitly what is required. Example:

* libvirtd running (see [some-section][some-section] for installation instructions)
* Permissions to access `/dev/kvm` (see [another-section][another-section] for more info)

File: docs/flatcar-linux/kvm-libvirt.md [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+* Terraform v0.11.x, [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct), and [terraform-provider-libvirt](https://github.com/dmacvicar/terraform-provider-libvirt) installed locally
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: Break into 3 separate lines?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+## Terraform Setup
+
+Install [Terraform](https://www.terraform.io/downloads.html) v0.11.x on your system.
+
+```sh
+$ terraform version
+Terraform v0.11.13
+```
+
+Add the [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct) plugin binary for your system to `~/.terraform.d/plugins/`, noting the final name.
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: Long line.

>noting the final name

What does "final" mean?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+Build terraform-provider-libvirt from source because pool definition feature was not released yet:
+
+```sh
+$ go get github.com/dmacvicar/terraform-provider-libvirt
+$ cd $GOPATH/src/github.com/dmacvicar/terraform-provider-libvirt
+$ GO111MODULE=on GOFLAGS=-mod=vendor make install
+$ cp $GOPATH/bin/terraform-provider-libvirt ~/.terraform.d/plugins/terraform-provider-libvirt
+```
+
+Read [concepts](/docs/architecture/concepts.md) to learn about Terraform, modules, and organizing resources. Change to your infrastructure repository (e.g. `infra`).
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: What is an "infrastructure repository"? I understand it's a directory but we should probably provide a better explanation as to where the user should create the `.tf` files.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+
+```sh
+cd infra/vmcluster
+```
+
+## Provider
+
+Create the `provider.tf` file:
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: I suggest "Create a file named `provider.tf`". The current phrasing assumes the user knows about this file.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+provider "libvirt" {
+  # TODO: add version once released
+  uri = "qemu:///system"
+  alias = "default"
+}
+```
+
+## Cluster
+
+Define a Kubernetes cluster using the module [kvm-libvirt/flatcar-linux/kubernetes](https://github.com/kinvolk/lokomotive-kubernetes/tree/master/kvm-libvirt/flatcar-linux/kubernetes).
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: Long line.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+
+Create the `mycluster.tf` file:
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: "Create a file named `mycluster.tf`"

File: docs/flatcar-linux/kvm-libvirt.md [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+
+```tf
+module "controller" {
+  source = "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes?ref=<hash>"
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: `ref=<hash>` will not work for users. They would have to edit it, which assumes they know about how Terraform modules work as well as the correct ref to use. Can we change this to a value that will work as is?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+
+  pool_name = "one"
+
+  count = 1
+
+  kubeconfig = "${module.controller.kubeconfig}"
+
+  labels = "node.supernova.io/role=backend,node-role.kubernetes.io/backend="
+}
+
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: Redundant empty line.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+```
+
+Now download the Flatcar image (example here is for Flatcar Edge) and increase its size as desired for your use case:
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: >example here is for Flatcar Edge

I suggest `the example below uses the "Edge" Flatcar channel`.

Comment (johananl) [reply docs/flatcar-linux/kvm-libvirt.md]: >and increase its size as desired for your use case

Why do I need to increase the size? What size should I increase to? Let's explain this and not assume the user knows about this.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+
+```sh
+$ wget https://edge.release.flatcar-linux.net/amd64-usr/2135.99.0/flatcar_production_qemu_image.img.bz2
```
Comment (johananl) [inline docs/flatcar-linux/kvm-libvirt.md]: `wget`, `bunzip2` and `qemu-img` are not listed as requirements so I expect these commands to fail for some users.

**Review by johananl — COMMENTED** [review]
@pothos I started reviewing this but haven't finished yet. I'm submitting what I currently have in the meantime and will do my best to finish testing this on Monday.

**Comment by johananl** [comment]
>Is this something we could use for the CI? We currently have another VM setup for that.

@blixtra our CI process uses the *bare-metal* target. This one adds a **new** target (libvirt).

File: docs/flatcar-linux/kvm-libvirt.md (hunk 8 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Thanks for the feedback. Yes, I've included your suggestions to make things more clear.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 16 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: I thought about adding the commit ID of the PR but that's a cyclic problem. So I just assume master will always work and added a note to add a specific commit if reproducability is needed.

**Comment by pothos** [comment]
I have to fix a problem with multiple nodes.

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+  name    = "${var.cluster_name}-controller-${count.index}-ignition"
+  pool    = "${libvirt_pool.volumetmp.name}"
+  count   = "${var.controller_count}"
+  content = "${element(data.ct_config.controller-ignitions.*.rendered, count.index)}"
+}
+
+resource "libvirt_domain" "controller-machine" {
+  count  = "${var.controller_count}"
+  name   = "${var.cluster_name}-controller-${count.index}"
+  vcpu   = "1"
```
Comment (invidian) [inline kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: I think those 2 parameters should be configurable.

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+  memory = "2048"
+
+  coreos_ignition = "${element(libvirt_ignition.ignition.*.id, count.index)}"
+
+  disk {
+    volume_id = "${element(libvirt_volume.controller-disk.*.id, count.index)}"
+  }
+
+  graphics {
```
Comment (invidian) [inline kvm-libvirt/flatcar-linux/kubernetes/controllers.tf:58]: We can disable GUI by default or make it configurable?

**Comment by invidian** [comment]
BTW I stripped the code a bit and it's actually also a cool way of having a bunch of Flatcar instances running locally :)

**Comment by pothos** [comment]
I think I've fixed the multi-node configuration count mistake.

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 21 above)
Comment (pothos) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf:58]: Haven't tried yet!

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 20 above)
Comment (pothos) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: Yes, done!

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 21 above)
Comment (pothos) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf:58]: I haven't tried it still because it's actually nice if the VM has a working console where it shows that it has booted and some other basic information even though you could login with ssh (assuming nothing broke the boot).

I also read there was a bug when it's disabled, but not sure.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 22]
```diff
@@ -0,0 +1,344 @@
...
+Controllers are provisioned to run an `etcd-member` peer and a `kubelet` service.
+Workers run just a `kubelet` service. A one-time [bootkube](https://github.com/kubernetes-incubator/bootkube)
+bootstrap schedules the `apiserver`, `scheduler`, `controller-manager`, and `coredns` on controllers
+and schedules `kube-proxy` and `calico` (or `flannel`) on every node.
+A generated `kubeconfig` provides `kubectl` access to the cluster.
+
+## Requirements
+
+* At least 4 GB of free RAM
+* Running libvirtd system service, see [libvirt Setup]
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: libvirt Setup is not a clickable link.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+
+Add the [terraform-provider-ct](https://github.com/poseidon/terraform-provider-ct) plugin binary for your system to `~/.terraform.d/plugins/`, noting the `_v0.3.1` suffix.
+
+```sh
+wget https://github.com/poseidon/terraform-provider-ct/releases/download/v0.3.1/terraform-provider-ct-v0.3.1-linux-amd64.tar.gz
+tar xzf terraform-provider-ct-v0.3.1-linux-amd64.tar.gz
+mv terraform-provider-ct-v0.3.1-linux-amd64/terraform-provider-ct ~/.terraform.d/plugins/terraform-provider-ct_v0.3.1
+```
+
+Build terraform-provider-libvirt from source because pool definition feature was not released yet.
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: Maybe rephrase with `Build terraform-provider-libvirt from source because pool definition feature has not been released yet.`
Is there a link that you can add for more information on this?
Maybe add a To do note so that as soon as the feature gets released we can update it?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+This means not only libvirt should be installed but also the development packets.
+
+```sh
+$ sudo dnf install libvirt-devel  # or sudo apt install libvirt-dev
+$ go get github.com/dmacvicar/terraform-provider-libvirt
+$ cd $GOPATH/src/github.com/dmacvicar/terraform-provider-libvirt
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: Maybe use `pushd` here and `popd` after everything is done? Because if you go next steps about making an `infra/vmcluster` directory, the user is still in the `terraform-provider-libvirt` directory

File: docs/flatcar-linux/kvm-libvirt.md [hunk 25]
```diff
@@ -0,0 +1,344 @@
...
+$ cp $GOPATH/bin/terraform-provider-libvirt ~/.terraform.d/plugins/terraform-provider-libvirt
+```
+
+Read [concepts](/docs/architecture/concepts.md) to learn about Terraform, modules, and organizing resources if the following confuses you.
+
+A Terrafrom project should be created in a new clean directory.
+Assuming there is a parent directory `infra` for all the terraform projects, we create a new subdirectory `vmcluster`.
+
+```sh
+mkdir infra/vmcluster
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]:
```
➜  mkdir infra/vmcluster
mkdir: cannot create directory ‘infra/vmcluster’: No such file or directory

➜  mkdir -p infra/vmcluster
```

But, why do we need to make two directory levels though? :confused:

File: docs/flatcar-linux/kvm-libvirt.md [hunk 26]
```diff
@@ -0,0 +1,344 @@
...
+$ sudo dnf install wget bzip2 qemu-img  # or sudo apt install wget bzip2 qemu-utils
+$ wget https://edge.release.flatcar-linux.net/amd64-usr/2135.99.0/flatcar_production_qemu_image.img.bz2
+$ bunzip2 flatcar_production_qemu_image.img.bz2
+$ qemu-img resize flatcar_production_qemu_image.img +5G
+```
+
+Now check the contents of `mycluster.tf`.
+Change at least the `ssh_keys` entry to match the content of your `~/.ssh/id_KEY.pub` and
+change the `os_image_unpacked` path to point to the image location from above.
+
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md:220]: I think this step of downloading the image should happen before the user defines the file. At first, I was confused about file path for `os_image_unpacked` as I thought it will be downloaded automatically.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+## ssh-agent
+
+Initial bootstrapping requires `bootkube.service` to be started on one controller node.
+Terraform uses `ssh-agent` to automate this step.
+Add your SSH private key to `ssh-agent` so that it's unlocked if secured with a passphrase
+(you can skip this if the key way created with the GNOME Passwords and Keys tool).
+
+```sh
+ssh-add ~/.ssh/id_KEY
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: You can leave it as `~/.ssh/id_rsa` since that is usually like a default.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 28]
```diff
@@ -0,0 +1,344 @@
...
+Reference the [variables docs](#variables) or the
+[variables.tf](https://github.com/kinvolk/lokomotive-kubernetes/blob/master/kvm-libvirt/flatcar-linux/kubernetes/variables.tf)
+source.
+
+## ssh-agent
+
+Initial bootstrapping requires `bootkube.service` to be started on one controller node.
+Terraform uses `ssh-agent` to automate this step.
+Add your SSH private key to `ssh-agent` so that it's unlocked if secured with a passphrase
+(you can skip this if the key way created with the GNOME Passwords and Keys tool).
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: s/way/was

File: docs/flatcar-linux/kvm-libvirt.md [hunk 29]
```diff
@@ -0,0 +1,344 @@
...
+
+Reference the [variables docs](#variables) or the
+[variables.tf](https://github.com/kinvolk/lokomotive-kubernetes/blob/master/kvm-libvirt/flatcar-linux/kubernetes/variables.tf)
+source.
+
+## ssh-agent
+
+Initial bootstrapping requires `bootkube.service` to be started on one controller node.
+Terraform uses `ssh-agent` to automate this step.
+Add your SSH private key to `ssh-agent` so that it's unlocked if secured with a passphrase
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md]: `so that it is unlocked if it was secured with a passphrase`?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 30]
```diff
@@ -0,0 +1,344 @@
...
+ssh-add ~/.ssh/id_KEY
+ssh-add -L
+```
+
+## Apply
+
+Initialize the config directory since this is the first use with Terraform.
+
+```sh
+terraform init
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md:243]:
```
 terraform init
Initializing modules...
- module.controller
  Getting source "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes"
Error downloading modules: Error loading modules: subdir "kvm-libvirt/flatcar-linux/kubernetes" not found
```
Pending when this PR gets merged, I think this readme should still reference the branch the source is coming from to make it easy for others trying to use this.
```
source = "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes?ref=kai/terraform-libvirt"
```
Same with this:
```
terraform init
Initializing modules...
- module.controller
  Getting source "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes?ref=kai/terraform-libvirt"
- module.worker-pool-one
  Getting source "git::https://github.com/kinvolk/lokomotive-kubernetes//kvm-libvirt/flatcar-linux/kubernetes/workers"
Error downloading modules: Error loading modules: subdir "kvm-libvirt/flatcar-linux/kubernetes/workers" not found
```
When it is eventually approved, it can be changed to reference the right value.

**Review by kosyfrances — COMMENTED** [review]
While trying to apply terraform, I got hit with 
```
Error: Error applying plan:

2 error(s) occurred:

* module.worker-pool-one.libvirt_domain.worker-machine: 1 error(s) occurred:

* libvirt_domain.worker-machine: Error creating libvirt domain: virError(Code=1, Domain=10, Message='internal error: process exited while connecting to monitor: 2019-06-20T11:05:59.605317Z qemu-system-x86_64: -drive file=/var/tmp/vmszqzs07/vmcluster-one-worker-0.qcow2,format=qcow2,if=none,id=drive-virtio-disk0: Could not open '/var/tmp/vmszqzs07/vmcluster-one-worker-0.qcow2': Permission denied')
* module.controller.libvirt_domain.controller-machine: 1 error(s) occurred:

* libvirt_domain.controller-machine: Error creating libvirt domain: virError(Code=1, Domain=10, Message='internal error: process exited while connecting to monitor: 2019-06-20T11:05:59.605666Z qemu-system-x86_64: -drive file=/var/tmp/vmszqzs07/vmcluster-controller-0.qcow2,format=qcow2,if=none,id=drive-virtio-disk0: Could not open '/var/tmp/vmszqzs07/vmcluster-controller-0.qcow2': Permission denied')
```

Left a couple of notes from my review so far. I will try setting up and continuing with review again some other time.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 22 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Thanks

File: docs/flatcar-linux/kvm-libvirt.md (hunk 23 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Changed this information, it's actually released now ;)

File: docs/flatcar-linux/kvm-libvirt.md (hunk 24 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Good point! Have removed this.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 25 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Simplified this ;)

File: docs/flatcar-linux/kvm-libvirt.md (hunk 26 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md:220]: Thanks, have moved it to the top.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 27 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Ok, I assume that those with another path can spot this themselves.

File: docs/flatcar-linux/kvm-libvirt.md (hunk 28 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Thanks :)

File: docs/flatcar-linux/kvm-libvirt.md (hunk 30 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md:243]: Good idea to use the branch name, thanks!

**Comment by pothos** [comment]
@kosyfrances Thanks for the review, can we have a look how to solve the error together?

File: docs/flatcar-linux/kvm-libvirt.md [hunk 31]
```diff
@@ -0,0 +1,349 @@
...
+Download the Flatcar image. The example below uses the "Edge" Flatcar channel.
+You need to increase its size depending on your use case because the initial size
+is only enough for some small pods to be loaded. Starting pods with normal-sized
+container images may already fail because the disk space will be filled.
+That's why the example below directly increases the image size by 5 GB to make
+sure that a couple of container images fit on the node.
+
+```sh
+$ sudo dnf install wget bzip2 qemu-img  # or sudo apt install wget bzip2 qemu-utils
+$ wget https://edge.release.flatcar-linux.net/amd64-usr/2135.99.0/flatcar_production_qemu_image.img.bz2
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: Maybe we can bundle all image-related step as `command1 && command2 && command3` so user just need to copy and execute one line?

I would also change the version to `current`, so user don't have to wonder which version is the latest.

File: docs/flatcar-linux/kvm-libvirt.md [hunk 32]
```diff
@@ -0,0 +1,349 @@
...
+```sh
+wget https://github.com/dmacvicar/terraform-provider-libvirt/releases/download/v0.5.2/terraform-provider-libvirt-0.5.2.Fedora_28.x86_64.tar.gz
+tar xzf terraform-provider-libvirt-0.5.2.Fedora_28.x86_64.tar.gz
+mv terraform-provider-libvirt ~/.terraform.d/plugins/terraform-provider-libvirt_v0.5.2
+```
+
+
+Read [concepts](/docs/architecture/concepts.md) to learn about Terraform, modules, and organizing resources if the following confuses you.
+
+A Terrafrom project should be created in a new clean directory.
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: `s/Terrafrom/Terraform`

File: docs/flatcar-linux/kvm-libvirt.md [hunk 33]
```diff
@@ -0,0 +1,349 @@
...
+  alias = "default"
+}
+```
+
+## Cluster
+
+Define a Kubernetes cluster using the module
+[kvm-libvirt/flatcar-linux/kubernetes](https://github.com/kinvolk/lokomotive-kubernetes/tree/master/kvm-libvirt/flatcar-linux/kubernetes).
+
+Create a file called `mycluster.tf` with these contents but remember to change the `CHANGEME` values:
```
Comment (invidian) [inline docs/flatcar-linux/kvm-libvirt.md]: Rather than relying on user reading docs, can we make CHANGEME values a required variables here?

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 21 above)
Comment (invidian) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf:58]: Fair enough :)

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+  network_interface {
+    network_id = "${libvirt_network.vmnet.id}"
+    hostname = "${var.cluster_name}-controller-${count.index}"
+    addresses = ["${cidrhost(var.node_ip_pool, 10 + count.index)}"]  # TODO: use as public addr in kubeconfig
+    wait_for_lease = true
+  }
+
+}
+
+resource "libvirt_network" "vmnet" {
```
Comment (invidian) [inline kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: I'd move this resource up (before `libvirt_domain`), to make structure of file match the order of execution.

**Review by invidian — CHANGES_REQUESTED** [review]
Just a few nits, but otherwise LGTM!

File: docs/flatcar-linux/kvm-libvirt.md [hunk 35]
```diff
@@ -0,0 +1,349 @@
...
+
+Add the [terraform-provider-libvirt](https://github.com/dmacvicar/terraform-provider-libvirt) plugin binary for your system
+to `~/.terraform.d/plugins/`, noting the `_v0.5.2` suffix.
+You should use the 0.5.2 release because it includes the experimental
+[pool definition feature](https://github.com/dmacvicar/terraform-provider-libvirt/commit/9f00f3d46c489f24c71d02fde816f5fda34d3f7c)
+but building from source is also an option.
+
+```sh
+wget https://github.com/dmacvicar/terraform-provider-libvirt/releases/download/v0.5.2/terraform-provider-libvirt-0.5.2.Fedora_28.x86_64.tar.gz
+tar xzf terraform-provider-libvirt-0.5.2.Fedora_28.x86_64.tar.gz
```
Comment (kosyfrances) [inline docs/flatcar-linux/kvm-libvirt.md:101]: Isn't this Fedora specific?

Comment (kosyfrances) [reply docs/flatcar-linux/kvm-libvirt.md:101]: Rather than Fedora maybe say Distro or tell the user that this is Fedora specific, referring them to the releases page where they can find their distro

File: docs/flatcar-linux/kvm-libvirt.md (hunk 31 above)
Comment (kosyfrances) [reply docs/flatcar-linux/kvm-libvirt.md]: I'll prefer we leave the commands as they are though.

Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md]: Thanks!

File: kvm-libvirt/flatcar-linux/kubernetes/controllers.tf (hunk 34 above)
Comment (pothos) [reply kvm-libvirt/flatcar-linux/kubernetes/controllers.tf]: Yes, done

File: docs/flatcar-linux/kvm-libvirt.md (hunk 35 above)
Comment (pothos) [reply docs/flatcar-linux/kvm-libvirt.md:101]: done

**Comment by pothos** [comment]
I've removed the `?ref=thisbranch` part of the terraform URLs to make it ready for merging.
